### PR TITLE
[Build] Use Tycho's pom-model build properties to reduce pom files and build with Tycho 2.7.4

### PR DESCRIPTION
**Pull request by HannesWell**
Using Tycho's pom-model build properties allows to further reduce the number of pom files required in m2e.
Furthermore the pom of o.e.m2e.core can be removed if its content is moved into a dedicated profile of the root pom that is activated automatically if the modello file is present (which is only the case for m2e.core).

@laeubi I extended the Tycho-Pomless wiki-page with a dedicated section how to define Maven properties inside the `build.properties` of a Project, which also provides an example and allows to link to it directly:
https://github.com/eclipse/tycho/wiki/Tycho-Pomless#define-properties

### CODE REVIEW

**Comment by HannesWell** [comment]
It looks like that for the `tycho-surefire-plugin` the configuration property `useUIThread` is not assigned with the value specified in the `build.properties`.
@laeubi is it possible that the indirection to define properties in the parent pom that are referenced in a plug-in configuration and overwritten with a `pom.model.property` does not work?

**Comment by laeubi** [comment]
I'm not sure what exactly is the question, but if you like to debug pomless it is sometimes useful to dump the pom:

https://github.com/eclipse/tycho/wiki/Tycho-Pomless#useful-tips

beside that, I think you are using here the system property name, it might be that you need to specify a property first in the parent and pass it to the surefire plugin configuration in the pluginMangement section.

File: org.eclipse.m2e.core.tests/build.properties [hunk 1]
```diff
@@ -3,3 +3,7 @@ output.. = bin/
 bin.includes = META-INF/,\
                resources/,\
                .
+# Maven Properties, see: https://github.com/eclipse/tycho/wiki/Tycho-Pomless#define-properties
+pom.model.property.testClass = org.eclipse.m2e.core.AllTests
+pom.model.property.tycho.surefire.useUIHarness = false
```
Comment (laeubi) [inline org.eclipse.m2e.core.tests/build.properties:8]: it *might be required to use
```suggestion
pom.model.property.useUIHarness = false
```

(same below)

**Comment by HannesWell** [comment]
> https://github.com/eclipse/tycho/wiki/Tycho-Pomless#useful-tips
> 
> beside that, I think you are using here the system property name, it might be that you need to specify a property first in the parent and pass it to the surefire plugin configuration in the pluginMangement section.

Thanks for the hint.

I found that the reason for the failure is a Bug in Tycho. Tycho-pomless is looking in the META-INF for the build.properties and not in the bundle root when enhancing the project properties.
I can implement a fix, but unfortunately not before Monday evening. It would bei great to postpone the 2.7.4 Release of Tycho so we can then use that fix soon.

**Comment by laeubi** [comment]
> I found that the reason for the failure is a Bug in Tycho. Tycho-pomless is looking in the META-INF for the build.properties and not in the bundle root when enhancing the project properties.

Thanks for looking into this, it seems we do not have yet any integration test covering that case.

> I can implement a fix, but unfortunately not before Monday evening. It would bei great to postpone the 2.7.4 Release of Tycho so we can then use that fix soon.

I'll postpone the release for one week, this would not harm much.

**Comment by HannesWell** [comment]
The test failures seem to be unrelated to this change and it looks like the pom.model properties are now picked up correctly. 🎉

**Comment by HannesWell** [comment]
Only due to the update to Tycho 2.7.4-SNAPSHOT three tests fail with the following exception. The root cause is at the bottom, the Maven processes launched to create Archetype projects fail. I'm in doubt that this is a regression of Tycho 2.7.4, since I cant see any change since 2.7.3 that is obviously related. I think it is more likely because a SNAPSHOT version is used.

```
org.eclipse.core.runtime.CoreException: Failed to create project.
	at org.eclipse.m2e.core.ui.internal.archetype.ArchetypeGenerator.createArchetypeProjects(ArchetypeGenerator.java:140)
	at org.eclipse.m2e.core.ui.internal.archetype.ArchetypeGenerator.createArchetypeProjects(ArchetypeGenerator.java:76)
	at org.eclipse.m2e.tests.common.AbstractMavenProjectTestCase.lambda$4(AbstractMavenProjectTestCase.java:707)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2380)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2405)
	at org.eclipse.m2e.tests.common.AbstractMavenProjectTestCase.createProjectsFromArchetype(AbstractMavenProjectTestCase.java:705)
	at org.eclipse.m2e.tests.common.AbstractMavenProjectTestCase.createProjectsFromArchetype(AbstractMavenProjectTestCase.java:699)
	at org.eclipse.m2e.tests.BuildPathManagerTest.testArchetypeProject(BuildPathManagerTest.java:1085)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray2(ReflectionUtils.java:206)
	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:161)
	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:84)
	at org.eclipse.tycho.surefire.osgibooter.OsgiSurefireBooter.run(OsgiSurefireBooter.java:128)
	at org.eclipse.tycho.surefire.osgibooter.AbstractUITestApplication$1.run(AbstractUITestApplication.java:37)
	at org.eclipse.e4.ui.internal.workbench.swt.E4Testable.lambda$1(E4Testable.java:127)
	at org.eclipse.swt.widgets.RunnableLock.run(RunnableLock.java:40)
	at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Synchronizer.java:132)
	at org.eclipse.swt.widgets.Display.runAsyncMessages(Display.java:5000)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4480)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1155)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1046)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:644)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:551)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:156)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:152)
	at org.eclipse.tycho.surefire.osgibooter.UITestApplication.runApplication(UITestApplication.java:29)
	at org.eclipse.tycho.surefire.osgibooter.AbstractUITestApplication.run(AbstractUITestApplication.java:122)
	at org.eclipse.tycho.surefire.osgibooter.UITestApplication.start(UITestApplication.java:35)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:136)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:402)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:659)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:596)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1467)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1440)
Caused by: java.util.concurrent.ExecutionException: java.lang.RuntimeException: Process exit value was 1
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at org.eclipse.m2e.core.ui.internal.archetype.ArchetypeGenerator.createArchetypeProjects(ArchetypeGenerator.java:125)
	... 74 more
Caused by: java.lang.RuntimeException: Process exit value was 1
	at org.eclipse.m2e.internal.launch.EclipseMavenLauncher$1.lambda$1(EclipseMavenLauncher.java:94)
	at java.base/java.util.Optional.ifPresent(Optional.java:178)
	at org.eclipse.m2e.internal.launch.EclipseMavenLauncher$1.launchesTerminated(EclipseMavenLauncher.java:88)
	at org.eclipse.debug.internal.core.LaunchManager$LaunchesNotifier.run(LaunchManager.java:305)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
	at org.eclipse.debug.internal.core.LaunchManager$LaunchesNotifier.notify(LaunchManager.java:262)
	at org.eclipse.debug.internal.core.LaunchManager.fireUpdate(LaunchManager.java:1038)
	at org.eclipse.debug.core.Launch.fireTerminate(Launch.java:479)
	at org.eclipse.debug.core.Launch.handleDebugEvents(Launch.java:644)
	at org.eclipse.debug.core.DebugPlugin$EventNotifier.run(DebugPlugin.java:1214)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
	at org.eclipse.debug.core.DebugPlugin$EventNotifier.dispatch(DebugPlugin.java:1248)
	at org.eclipse.debug.core.DebugPlugin$EventDispatchJob.run(DebugPlugin.java:480)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
```

**Comment by HannesWell** [comment]
Great it finally succeeded 🎉

> Only due to the update to Tycho 2.7.4-SNAPSHOT three tests fail with the following exception. The root cause is at the bottom, the Maven processes launched to create Archetype projects fail. I'm in doubt that this is a regression of Tycho 2.7.4, since I cant see any change since 2.7.3 that is obviously related. I think it is more likely because a SNAPSHOT version is used.

Looks like it is indeed somehow related to using SNAPSHOT versions of Tycho.
Anyway this can be merged now. 😃

**Comment by laeubi** [comment]
> Looks like it is indeed somehow related to using SNAPSHOT versions of Tycho.

The problem is that the test uses some strange kind of remote repository configs, tycho snapshots are most probably not included there...

**Comment by HannesWell** [comment]
Understand. So as long as we don't use SNAPSHOTs that not a big issue and if we do, we should keep that in mind that we don't have to worry.

**Pull request merged**